src/jit/flowgraph.cpp

```diff
@@ -21185,12 +20986,234 @@ void Compiler::fgDebugCheckBBlist(bool checkBBNum /* = false */, bool checkBBRef
+        BasicBlock* blockPred = pred->flBlock;
+        assert(blockPred->bbTraversalStamp == curTraversalStamp);
+
+        EHblkDsc* ehTryDsc = ehGetBlockTryDsc(block);
+        if (ehTryDsc != nullptr)
+        {
+            // You can jump to the start of a try
+            if (ehTryDsc->ebdTryBeg == block)
+            {
+                goto CHECK_EXCEPTION_HANDLER;
+            }
+
+            // You can jump within the same try region
+            if (bbInTryRegions(block->getTryIndex(), blockPred))
+            {
+                goto CHECK_EXCEPTION_HANDLER;
+            }
+
+            // The catch block can jump back into the middle of the try
+            if (bbInCatchHandlerRegions(block, blockPred))
+            {
+                goto CHECK_EXCEPTION_HANDLER;
+            }
+
+            // The end of a finally region is a BBJ_EHFINALLYRET block (during importing, BBJ_LEAVE) which
+            // is marked as "returning" to the BBJ_ALWAYS block following the BBJ_CALLFINALLY
+            // block that does a local call to the finally. This BBJ_ALWAYS is within
+            // the try region protected by the finally (for x86, ARM), but that's ok.
+            BasicBlock* prevBlock = block->bbPrev;
+            if (prevBlock->bbJumpKind == BBJ_CALLFINALLY && block->bbJumpKind == BBJ_ALWAYS &&
+                blockPred->bbJumpKind == BBJ_EHFINALLYRET)
+            {
+                goto CHECK_EXCEPTION_HANDLER;
+            }
+
+            printf("Jump into the middle of try region: BB%02u branches to BB%02u\n", blockPred->bbNum, block->bbNum);
+            assert(!"Jump into middle of try region");
+        }
+
+    CHECK_EXCEPTION_HANDLER:
+
+        EHblkDsc* ehHndDsc = ehGetBlockHndDsc(block);
+        if (ehHndDsc != nullptr)
+        {
+            // You can do a BBJ_EHFINALLYRET or BBJ_EHFILTERRET into a handler region
+            if ((blockPred->bbJumpKind == BBJ_EHFINALLYRET) || (blockPred->bbJumpKind == BBJ_EHFILTERRET))
+            {
+                goto CHECK_JUMP;
+            }
+
+            // Our try block can call our finally block
+            if ((block->bbCatchTyp == BBCT_FINALLY) && (blockPred->bbJumpKind == BBJ_CALLFINALLY) &&
+                ehCallFinallyInCorrectRegion(blockPred, block->getHndIndex()))
+            {
+                goto CHECK_JUMP;
+            }
+
+            // You can jump within the same handler region
+            if (bbInHandlerRegions(block->getHndIndex(), blockPred))
+            {
+                goto CHECK_JUMP;
+            }
+
+            // A filter can jump to the start of the filter handler
+            if (ehHndDsc->HasFilter())
+            {
+                goto CHECK_JUMP;
+            }
+
+            printf("Jump into the middle of handler region: BB%02u branches to BB%02u\n", blockPred->bbNum,
+                   block->bbNum);
+            assert(!"Jump into the middle of handler region");
+        }
+
+    CHECK_JUMP:
```

Comment:
Maybe this should be a separate function - `IsSuccessor` or something like that.